File: SEARCH_CPP/Interpolation_search.cpp
#include <iostream>


template <typename Type>
int64_t len(Type *begin, Type *end) {
	Type *cur = begin;
	int64_t count = 0;
	if (begin == end) { return 1; }
	while (cur != end) {
		count++;
		cur++;
	}
	return count;
}

/*******Интерполяционный поиск - interpolation_search********/
template <typename Type>
int64_t interpolation_search(Type *first, Type *last, const Type &elm) {
	
	int64_t mid;
	int64_t low = 0;
	int64_t high = len(first, last) - 1;

    while (low <= high && elm >= first[low] && elm <= first[high]) {
        if (low == high) {
            if (first[low] == elm) {
                return low;
            }
            return -1;
        }
        int64_t pos = low + ((static_cast<int64_t>(high - low) /
            (first[high] - first[low])) * (elm - first[low]));
        if (first[pos] == elm) {
            return pos;
        }
        if (first[pos] < elm) {
            low = pos + 1;
        }
        else {
            high = pos - 1;
        }
    }
    return -1;
}
/************************************************************/

// Рекурсивынй вариант
template <typename Type>
size_t interpolation_search_rec(Type *arr, const Type &value, size_t begin = 0, size_t end = 0) {

    size_t pos;
    if (begin <= end && value >= arr[begin] && value <= arr[end]) {
        pos = begin + (((double)(end - begin) / (arr[end] - arr[begin])) * (value - arr[begin]));

        if (arr[pos] == value)
            return pos;
        if (arr[pos] < value)
            return interpolation_search_rec(arr, value, pos + 1, end);
        if (arr[pos] > value)
            return interpolation_search_rec(arr, value, begin, pos - 1);
    }
    return -1;
}
/************************************************************/

int main() {

    int arr[] = { 10, 12, 13, 16, 18 };
    int n = sizeof(arr) / sizeof(arr[0]);

    std::cout << interpolation_search(arr, arr + n, 18)      << "\n";
    std::cout << interpolation_search_rec(arr, 16, 0, n - 1) << "\n";

    return 0;
}

File: SEARCH_PY/Interpolation_search.py
import random

"""*******Интерполяционный поиск - interpolation_search********"""
def interpolation_search(array, elem):
    high = len(array) - 1
    low = 0

    while low <= high and elem >= array[low] and elem <= array[high]:
        if low == high:
            if array[low] == elem:
                return low
            return -1
        pos = low + ((high - low) // (array[high] - array[low]) * (elem - array[low]))

        if array[pos] == elem:
            return pos
        if array[pos] < elem:
            low = pos + 1
        else:
            high = pos - 1
    return -1
"""************************************************************"""

#*********--Рекурсивный вариант--*******
def interpolation_search_rec(array, value, begin=0, end=0):
    if begin <= end and value >= array[begin] and value <= array[end]:
        pos = begin + ((end - begin) // (array[end] - array[begin])) * (value - array[begin])

        if array[pos] == value:
            return pos
        if array[pos] < value:
            return interpolation_search_rec(array, value, pos + 1, end)
        if array[pos] > value:
            return interpolation_search_rec(array, value, begin, pos - 1)
    return -1


def main():
    arr = [random.randint(-100, 100) for i in range(200)]
    arr.sort()
    id = interpolation_search(arr, 10)
    id2 = interpolation_search_rec(arr, 10, 0, len(arr) - 1)
    print(f"id = {id}\nid2 = {id2}")
    if id != -1:
        print(arr[id])
    if id2 != -1:
        print(arr[id2])


if __name__ == "__main__":
    main()